extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/WrongAnnotationUsageProcessor.java

```diff
@@ -72,35 +75,72 @@ void detect(ArcConfig config, ApplicationIndexBuildItem applicationIndex, Custom
-                }
-                if (scopeAnnotations.isScopeIn(classAnnotations)) {
+                // Annotations declared on the class level, incl. the annotations added via transformers
+                List<AnnotationInstance> classLevelAnnotations = transformedAnnotations.getAnnotations(clazz).stream()
+                        .filter(a -> a.target().kind() == AnnotationTarget.Kind.CLASS).collect(Collectors.toList());
+                if (scopeAnnotations.isScopeIn(classLevelAnnotations)) {
                     validationErrors.produce(new ValidationErrorBuildItem(
                             new IllegalStateException(String.format(
                                     "The %s class %s has a scope annotation but it must be ignored per the CDI rules",
                                     clazz.nestingType().toString(), clazz.name().toString()))));
-                } else if (clazz.annotationsMap().containsKey(DotNames.OBSERVES)) {
-                    validationErrors.produce(new ValidationErrorBuildItem(
-                            new IllegalStateException(String.format(
-                                    "The %s class %s declares an observer method but it must be ignored per the CDI rules",
-                                    clazz.nestingType().toString(), clazz.name().toString()))));
-                } else if (clazz.annotationsMap().containsKey(DotNames.PRODUCES)) {
-                    validationErrors.produce(new ValidationErrorBuildItem(
-                            new IllegalStateException(String.format(
-                                    "The %s class %s declares a producer but it must be ignored per the CDI rules",
-                                    clazz.nestingType().toString(), clazz.name().toString()))));
-                } else if (Annotations.containsAny(classAnnotations, interceptorResolverBuildItem.getInterceptorBindings())
-                        || Annotations.containsAny(clazz.annotations(),
-                                interceptorResolverBuildItem.getInterceptorBindings())) {
-                    // detect interceptor bindings on nested classes
+                } else if (Annotations.containsAny(classLevelAnnotations,
+                        interceptorResolverBuildItem.getInterceptorBindings())) {
+                    // detect interceptor bindings declared at nested class level
                     validationErrors.produce(new ValidationErrorBuildItem(
                             new IllegalStateException(String.format(
                                     "The %s class %s declares an interceptor binding but it must be ignored per CDI rules",
                                     clazz.nestingType().toString(), clazz.name().toString()))));
                 }
+
+                // iterate over methods and verify those
+                // since JDK 16, you can have static method inside inner non-static class - skip those
+                for (MethodInfo methodInfo : clazz.methods()) {
```

Comment:
Hm, so previously we only detected annotations declared on the class and now we also detect annotations added through transformers, right? I think that the previous check was good enough and it could be reimplemented so that you obtain all annotations from the `clazz.annotationsMap()`, find those that target methods and skip `static` methods. That could be much more efficient but if you think it's important to include transformers then we should maybe also think about inherited methods ;-).

Reply:
We don't have a test for either so I'm not quite sure about the word "support"...

Reply:
In that case, I propose we limit that to interceptors (due to a historical accident, it certainly wasn't intentional, was it?), and reject everything else.

Reply:
I definitely don't think it's worth the hassle to delve into inherited methods...

> Hm, so previously we only detected annotations declared on the class...

Not really, for observers and producers you are right - for bindings we actually checked both, see [this code](https://github.com/quarkusio/quarkus/blob/main/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/WrongAnnotationUsageProcessor.java#L95-L97).
I decided to use transformed versions wherever applicable as that makes to code more consistent and foolproof.
However, if you are truly so much against it @mkouba, I could do it the way you suggest :)

Reply:
> In that case, I propose we limit that to interceptors (due to a historical accident, it certainly wasn't intentional, was it?), and reject everything else.

What historical accident do you have in mind??

Reply:
> However, if you are truly so much against it @mkouba, I could do it the way you suggest :)

I'm not against it. It's more like _"ok, this seems to be a little bit more correct but does not bring any real benefit"_ ;-)

Reply:
> > In that case, I propose we limit that to interceptors (due to a historical accident, it certainly wasn't intentional, was it?), and reject everything else.
> 
> What historical accident do you have in mind??

When adding support for `static` methods interception, we certainly didn't have `static` methods on inner classes in mind (because they didn't exist yet). It is a historical accident that interception of `static` methods on inner classes used to work (and then we broke it).

I suggest we keep that case working, but still validate `static` methods on inner classes for observers and producers, because we never intended for those to work, IIUC?

Reply:
> In that case, I propose we limit that to interceptors (due to a historical accident, it certainly wasn't intentional, was it?), and reject everything else.

Supporting static interceptors was intentional (the fact that it now works in this setup is an accident though).
I am also unsure what the status of observers and producers is so you're right we can keep those as it was.
I'll modify the code.

Reply:
> When adding support for static methods interception, we certainly didn't have static methods on inner classes in mind (because they didn't exist yet). It is a historical accident that interception of static methods on inner classes used to work (and then we broke it).

Hm, but the same would apply to observers and producers... except that static producers/observers declared on inner classes never worked because we only find the producers/observers declared on bean classes and inner classes are not considered beans.

So +1 for your proposal!

Reply:
> Supporting static interceptors was intentional (the fact that it now works in this setup is an accident though).

I think that Ladislav meant _"interception of static method declared on inner classes was not intentional"_ ;-)

Reply:
Yeah, exactly. Sorry, it's Friday and my communication skills are even worse than usual :-)